### DISCUSSION TITLE
Sensor with unit of measure will not display as a line graph in history

### DISCUSSION THREAD

**Discussion by ghost** [make-all/tuya-local#2564] [Q&A]
Hello,
I am pretty new to Home Assitant an i try to integrate a powerworld heatpump. I had a  nice start from other models integrations from other members. But unfortunatly not al sensors wil display as a line graph in the history. I googled and found some tips but none of them is working. If make it a number then it display's as a linegraph, But as a sensor it is not working. Also i added unit_of _mesurment and still only the colored bar.

Can somebody help me out? (see the Yaml code below and the picture of the colored bar)
![Schermafbeelding 2024-11-27 141155](https://github.com/user-attachments/assets/617f8734-86b4-4561-9756-6b3b4fb0e68a)

![image](https://github.com/user-attachments/assets/2dd18661-5d26-4636-9dbe-c5a767038d1d)
  - entity: sensor
    name: Coefficient of performance
    category: diagnostic
    dps:
      - id: 140
        type: base64
        optional: true
        name: sensor
        unit_of_measurement: ppm
        mapping:
          - mask: "0000000000000000000000000000000000000000FFFFFFFF\
                   000000000000000000000000000000000000000000000000\
                   000000000000000000000000000000000000000000000000\
                   0000000000000000"
            scale: 10
        range:
          min: 0
          max: 7

**Comment by make-all** [comment]
This is because the source of your sensor comes from a string, not a number. It is a problem that your device manufacturer has chosen to encode data in a complex way like this, and better handling for manufacturers' bad decisions is not something that I am going to prioritise for this integration.

**Reply by ghost** [reply]
Okay, thanks for the fast reply.
Is it possible for a workaround by me? Because if i make it a number the line graph is displayed. Some string to number conversion is there, do yuo have tips for me or have just to live with it? 
(conversion function of the string to a number when reading the data)

**Reply by make-all** [reply]
If you use a template sensor, you can probably turn it into a proper number (using a float or int filter).

**Reply by ghost** [reply]
thanks, i had a quick look at the template sensor. Looks very difficult for me, i gonna try it.

**Reply by ghost** [reply]
p.s. what also is a possible solution, and i think a simpel soluction. When displaying the number data with the readonly flag set. Do not display the input field or slider. Just display it conform the sensor data. The history data is then displayed as a line graph.

**Reply by ghost** [reply]
hello,
making a template sensor did the trick.
i also made a check for correct value's but that seems noet to work. see the code below. 
Do you have a tip for me why. I checked the code in de template editor and ther its working.
see the yaml code below:
======================
- sensor:
    - name: COP_as_number
      state: "{{ states['sensor.heat_pump_coefficient_of_performance']['state'] }}"
      unit_of_measurement: " "
      attributes:
        value_template: >
          {% set raw_value = states('sensor.heat_pump_coefficient_of_performance') %}
          {% if raw_value | is_number  %}
            {% if raw_value | float(default=0) >= 0.0 and raw_value | float(default=0) < 7.0 %}  
              {{ raw_value | float(0) }}
            {% else %}
              0.0
            {% endif %}
          {% else %}
            0.0
          {% endif %

**Reply by ghost** [reply]
a found availability and that is working!

**Reply by ghost** [reply]
Hello Jason,
Huubkuh here again, i managed to get it working with your tip to use a template sensor (thanks for that) and also made a specific device yaml file for my poolworld 58411 hetapump en the devices directory.
It 'was' working okay.... But after i have run the Tuya Local update my device file was gone and the heatpump was not seen bij Home Assistant anymore. It seems that the directory with device yaml files is completly deleted and replaced by the device files in the update. When i uploaded my device yaml file from a backup i could not get it working again. A restore of a full Home Asssitant backup made it working again but i lost historical data with it.
What did i wrong? 
Is there a way to let i working after an update?
Thanks in advance, Huubkuh
